{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c4c5320-dd9d-4a6e-82f2-aa1f88ed59d8",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Clinical Analysis: Prostate Tumor Stage at T2b"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a668c51-86e0-4933-b6dc-70454d529a2c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Loading the Dataset file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "529b0662-1999-493c-9d9b-583a72e9436d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np \n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "#list(data.columns)\n",
    "list(data[\"TUMOR_STAGE_T2b\"].unique())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2d0814b-4983-495a-ae9b-8fa6b656940c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Test for Normal Distribution and Homogeneity of Variance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f57c03f-cdc4-4a46-ae12-1f7218d4d541",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "################## QQ plots with Shapiro Wilk Test Annotations:\n",
    "\n",
    "# Load the data\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "# Define function to plot QQ plots with Shapiro Wilk Test Annotations\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import statsmodels.api as sm\n",
    "import scipy.stats as stats\n",
    "\n",
    "def analyze_gene(data, gene_name, hue_by, condition_order):\n",
    "    sns.set(style=\"whitegrid\")\n",
    "    fig, axes = plt.subplots(nrows=2, ncols=3, figsize=(12, 8))\n",
    "    axes = axes.flatten()\n",
    "    for idx, condition in enumerate(condition_order):\n",
    "        ax = axes[idx]\n",
    "        subset_data = data[data[hue_by] == condition]\n",
    "        # Check if the subset data has at least 3 data points\n",
    "        if len(subset_data[gene_name]) < 3:\n",
    "            ax.set_title(f'Not enough data for {gene_name}\\n{condition}')\n",
    "            continue\n",
    "        sm.qqplot(subset_data[gene_name], line='s', ax=ax)\n",
    "        ax.set_title(f'QQ Plot\\n{condition}\\n{gene_name}')\n",
    "        # Perform Shapiro-Wilk test\n",
    "        shapiro_test_stat, shapiro_p_value = stats.shapiro(subset_data[gene_name])\n",
    "        color = 'red' if shapiro_p_value < 0.05 else 'green'\n",
    "        ax.text(0.5, 0.1, f'S-W p-value: {shapiro_p_value:.4f}', transform=ax.transAxes, fontsize=10, color=color)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'{gene_name}_QQPLOT.png', format='png', dpi=600, bbox_inches='tight')\n",
    "    plt.show()\n",
    "\n",
    "# Specify the genes you want to analyze\n",
    "genes_to_analyze = ['MIR17_EXP', 'MIR20B_EXP', 'MIR106A_EXP']  # Add more genes as needed\n",
    "#Specify the phenotype you want to analyze\n",
    "hue_by = 'TUMOR_STAGE_T2b_CONDITION'\n",
    "# Specify the order of sample conditions\n",
    "condition_order = ['PRE-RT, ≤T2b', '4WKSPD1, ≤T2b', '8WKSPD1, ≤T2b', \n",
    "                   'PRE-RT, >T2b', '4WKSPD1, >T2b', '8WKSPD1, >T2b']\n",
    "# Call the function for each gene\n",
    "for gene in genes_to_analyze:\n",
    "    analyze_gene(data, gene, hue_by, condition_order)\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63c89c47-6a3d-458d-8b1a-d5d9e45032f1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "###################### Levene Test for Homogeneity of Variance:\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy.stats import levene\n",
    "\n",
    "def run_levene_test(dataframe, gene_to_analyze, groups_column, timepoints_column):\n",
    "    timepoints = np.unique(dataframe[timepoints_column])\n",
    "    groups = np.unique(dataframe[groups_column])\n",
    "    data_levene = []\n",
    "    for tp in timepoints:\n",
    "        for group in groups:\n",
    "            subset = dataframe[(dataframe[timepoints_column] == tp) & (dataframe[groups_column] == group)][gene_to_analyze].values\n",
    "            data_levene.append(subset)\n",
    "    statistic, p_values = levene(*data_levene)\n",
    "    print(f'Test Statistics: {round(statistic, 3)}')\n",
    "    print(f'p-values: {round(p_values, 3)}')\n",
    "\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "mirnas = [\"MIR17_EXP\", \"MIR20B_EXP\", \"MIR106A_EXP\"]\n",
    "\n",
    "for mirna in mirnas:\n",
    "    print(f'Levene Test for {mirna}')\n",
    "    run_levene_test(dataframe = data, gene_to_analyze = mirna, groups_column = \"TUMOR_STAGE_T2b\", timepoints_column = \"TIME_POINT\")\n",
    "    print(\"\\n\")\n",
    "    \n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52b23205-2a56-4946-ade0-4a7129645105",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Descriptive Statistics for the Gene Expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd359544-2692-4085-9144-d94f292a2e1c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "################### Obtaining Descr. Stat. on Clinical Outcome:\n",
    "\n",
    "# Load the data\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "from datetime import datetime\n",
    "\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "current_date = datetime.now().strftime('%Y_%m_%d')\n",
    "column = \"TUMOR_STAGE_T2b\"\n",
    "\n",
    "# Perform descriptive statistics on the data frame\n",
    "miRNA_exp_descr_stats = data.groupby(['TIME_POINT', column], sort=False)[['MIR17_EXP', 'MIR20B_EXP', 'MIR106A_EXP']].describe()\n",
    "# Round the values to a specific number of decimals\n",
    "miRNA_exp_descr_stats = miRNA_exp_descr_stats.round(decimals=3)\n",
    "# Reset index to make the DataFrame more readable in the output\n",
    "miRNA_exp_descr_stats.reset_index(inplace=True)\n",
    "outputfile_name = f'{current_date}_{column}_DESC_STATS.xlsx'\n",
    "# Save the result to a CSV file\n",
    "miRNA_exp_descr_stats.to_excel(outputfile_name, index=True)\n",
    "### Download and format in excel independently\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e9eeda1-646a-4a4c-9153-62e07f6d98ea",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Running Mann-Whitney U Test Manually "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76e21f08-819b-40be-9c1b-48e353fa5402",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "################################### Mann Whitney U Test Results\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "from scipy.stats import mannwhitneyu\n",
    "\n",
    "# Import the Data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "genes = [\"MIR17_EXP\", \"MIR20B_EXP\", \"MIR106A_EXP\"]\n",
    "timepoints = [\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"]\n",
    "column1 = \"TIME_POINT\"\n",
    "column2 = \"TUMOR_STAGE_T2b\"\n",
    "groups = data[column2].unique()\n",
    "\n",
    "# Subset the data for Mann-Whitney U Testing\n",
    "tuple_mir17 = []\n",
    "tuple_mir20b = []\n",
    "tuple_mir106a = []\n",
    "        \n",
    "for gene in genes:    \n",
    "    for tp in timepoints:\n",
    "        subset1 = data[(data[column1] == f'{tp}') & (data[column2] == f'{groups[0]}')][gene].values\n",
    "        subset2 = data[(data[column1] == f'{tp}') & (data[column2] == f'{groups[1]}')][gene].values\n",
    "        mypair = (subset1, subset2)\n",
    "        if gene == \"MIR17_EXP\":\n",
    "            tuple_mir17.append(mypair)\n",
    "        elif gene == \"MIR20B_EXP\":\n",
    "            tuple_mir20b.append(mypair)\n",
    "        else:\n",
    "            tuple_mir106a.append(mypair)\n",
    "\n",
    "### Accessing the Tuple\n",
    "# tuple_miRNA[0-2], provides the timepoints, 0=prert, 1=4wkspd1, 2=8wkspd1\n",
    "# tuple_miRNA[0-2][0-1], provides the group [0='≤T2b', 1='>T2b']\n",
    "\n",
    "# Perform the Mann-Whitney U Test:\n",
    "for miRNA in genes:\n",
    "    print(f'Mann-Whitney U Test for {miRNA}')\n",
    "    if miRNA == \"MIR17_EXP\":\n",
    "        for idx, (group1_data, group2_data) in enumerate(tuple_mir17, start = 1):\n",
    "            if idx == 1:\n",
    "                print(f'Comparison at PRE-PT')\n",
    "            elif idx == 2:\n",
    "                print(f'Comparison at 4WKSPD1')\n",
    "            else: \n",
    "                print(f'Comparison at 8WKSPD1')\n",
    "            mannwhit_stat, p_value = mannwhitneyu(group1_data, group2_data, alternative = \"two-sided\")\n",
    "            print(f\"Mann-Whitney U Statistic: {mannwhit_stat:.3f}\")\n",
    "            print(f\"P-value: {p_value:.3f}\")\n",
    "            if p_value < 0.05:\n",
    "                print(f'Gene Expression is DIFFERENT')\n",
    "            else:\n",
    "                print(f'Gene Expression is NOT different')\n",
    "            print(\"\\n\")\n",
    "        print(\"\\n\"*3)\n",
    "    elif miRNA == \"MIR20B_EXP\":\n",
    "        for idx, (group1_data, group2_data) in enumerate(tuple_mir20b, start = 1):\n",
    "            if idx == 1:\n",
    "                print(f'Comparison at PRE-PT')\n",
    "            elif idx == 2:\n",
    "                print(f'Comparison at 4WKSPD1')\n",
    "            else: \n",
    "                print(f'Comparison at 8WKSPD1')\n",
    "            mannwhit_stat, p_value = mannwhitneyu(group1_data, group2_data, alternative = \"two-sided\")\n",
    "            print(f\"Mann-Whitney U Statistic: {mannwhit_stat:.3f}\")\n",
    "            print(f\"P-value: {p_value:.3f}\")\n",
    "            if p_value < 0.05:\n",
    "                print(f'Gene Expression is DIFFERENT')\n",
    "            else:\n",
    "                print(f'Gene Expression is NOT different')\n",
    "            print(\"\\n\")\n",
    "        print(\"\\n\"*3)\n",
    "    else:\n",
    "        for idx, (group1_data, group2_data) in enumerate(tuple_mir106a, start = 1):\n",
    "            if idx == 1:\n",
    "                print(f'Comparison at PRE-PT')\n",
    "            elif idx == 2:\n",
    "                print(f'Comparison at 4WKSPD1')\n",
    "            else: \n",
    "                print(f'Comparison at 8WKSPD1')\n",
    "            mannwhit_stat, p_value = mannwhitneyu(group1_data, group2_data, alternative = \"two-sided\")\n",
    "            print(f\"Mann-Whitney U Statistic: {mannwhit_stat:.3f}\")\n",
    "            print(f\"P-value: {p_value:.3f}\")\n",
    "            if p_value < 0.05:\n",
    "                print(f'Gene Expression is DIFFERENT')\n",
    "            else:\n",
    "                print(f'Gene Expression is NOT different')\n",
    "            print(\"\\n\")\n",
    "    \n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb433e08-d9dd-4d9a-8d07-ec6d82fae806",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plots stratified according to Tumor Stage Group with Mann-Whitney U Test Results Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13e36553-b8e7-48eb-8522-5cabe1844ab5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#################################### Mann-Whitney U Test Plots:\n",
    "\n",
    "# Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import product\n",
    "from statannotations.Annotator import Annotator\n",
    "\n",
    "# Define function to plot Mann-Whitney U Test Plots\n",
    "def plot_gene(x, data, hue, gene_name, ax, hue_order, custom_palette, PROPS, flierprops, stat_pairs):\n",
    "    sns.set(style=\"whitegrid\")\n",
    "    # Plot Boxplot\n",
    "    sns.boxplot(x=x, y=gene_name, data=data, order=[\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"], hue=hue, hue_order=hue_order,\n",
    "                palette=custom_palette, showmeans=False, showfliers=False,\n",
    "                meanprops={\"marker\": \"s\", \"markerfacecolor\": \"black\", \"markersize\": 5, \"markeredgecolor\": \"black\"},\n",
    "                linewidth=1.5, medianprops=dict(color=\"black\"), flierprops=flierprops, **PROPS, ax=ax)\n",
    "\n",
    "    # Get the legend from just the box plot\n",
    "    handles, labels = ax.get_legend_handles_labels()\n",
    "\n",
    "    sns.swarmplot(x=x, y=gene_name, data=data, order=[\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"], hue=hue, hue_order=hue_order,\n",
    "                  palette=custom_palette, dodge=True, size=5, linewidth=0.5, edgecolor=\"black\", ax=ax)\n",
    "\n",
    "    # Add Annotations\n",
    "    annotator = Annotator(ax=ax, pairs=stat_pairs, data=data, x=x, y=gene_name, order=[\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"],\n",
    "                          hue=hue, hue_order=hue_order, palette=custom_palette)\n",
    "    \n",
    "    annotator.configure(test=\"Mann-Whitney\").apply_and_annotate() # comparisons_correction=\"Bonferroni\"\n",
    "\n",
    "    # Remove the old legend\n",
    "    ax.legend_.remove()\n",
    "\n",
    "    # Add just the handles/labels from the box plot back\n",
    "    legend = ax.legend(handles, labels, loc=0)\n",
    "    legend.set_title(\"PCa Tumor Stage\", prop={\"size\": 9, \"family\": \"serif\"})\n",
    "    for text in legend.texts:\n",
    "        text.set_fontsize(9)  # Set the desired font size\n",
    "        text.set_family(\"serif\")  # Set the desired font family\n",
    "\n",
    "    # Map and update the labels to new ones\n",
    "    new_labels = {'≤T2b': \"≤T2b\", '>T2b': \">T2b\"}\n",
    "    for text, handle in zip(legend.texts, legend.legendHandles):\n",
    "        text.set_text(new_labels[text.get_text()])\n",
    "\n",
    "    labels = [item.get_text() for item in ax.get_xticklabels()]\n",
    "    labels[0] = 'PRE-PT'\n",
    "    labels[1] = '4WKSPD1'\n",
    "    labels[2] = '8WKSPD1'\n",
    "    ax.set_xticklabels(labels, family=\"serif\")\n",
    "    ax.set_yticklabels(ax.get_yticklabels(), family=\"serif\")\n",
    "\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.xaxis.tick_bottom()\n",
    "    ax.tick_params(axis='x', labelsize=9, labelrotation=0)\n",
    "    ax.tick_params(axis='y', labelsize=9)\n",
    "\n",
    "    ax.set_xlabel('Time Point', family=\"serif\", fontsize=12)\n",
    "    ax.set_ylabel('Gene Expression Fold Change', family=\"serif\", fontsize=12)\n",
    "    variable = gene_name\n",
    "    result = variable.split('_')[0]\n",
    "    ax.set_title(result, family=\"serif\")\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "608e79e5-0104-4f50-9664-43da4f410294",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from statannotations.Annotator import Annotator\n",
    "\n",
    "# Call the function to perform the plots \n",
    "# Load the data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "# Define the Pairs needed for the plot annotations with STATANNOTATOR\n",
    "mannwhitu_testpairs = [[(\"PRE-RT\", \"≤T2b\"), (\"PRE-RT\", \">T2b\")],\n",
    "                       [(\"4WKSPD1\", \"≤T2b\"), (\"4WKSPD1\", \">T2b\")],\n",
    "                       [(\"8WKSPD1\", \"≤T2b\"), (\"8WKSPD1\", \">T2b\")]]\n",
    "\n",
    "# Call Plotting Parameters\n",
    "column1 = \"TIME_POINT\"\n",
    "column2 = \"TUMOR_STAGE_T2b\"\n",
    "group_hue_order = [\"≤T2b\", \">T2b\"]\n",
    "group_custom_palette = {\"≤T2b\" : \"#E8E8E8\", \">T2b\": \"#505050\"}\n",
    "group_PROPS = {'boxprops': {'edgecolor': 'black'}, 'whiskerprops': {'color': 'black'}, 'capprops': {'color': 'black'}}\n",
    "group_flierprops = dict(markerfacecolor='grey', markersize=5, markeredgecolor='black')\n",
    "\n",
    "# Begin Plotting\n",
    "sns.set(style=\"whitegrid\")\n",
    "fig, axs = plt.subplots(nrows=1, ncols=3, figsize=(30, 8), dpi=600)\n",
    "# Call the function for each gene\n",
    "genes_to_plot = ['MIR17_EXP', 'MIR20B_EXP', 'MIR106A_EXP']  # Add more genes as needed\n",
    "for idx, gene in enumerate(genes_to_plot):\n",
    "    plot_gene(x=column1, data=data, gene_name=gene, ax=axs[idx], hue = column2, hue_order=group_hue_order,\n",
    "              custom_palette=group_custom_palette, PROPS=group_PROPS, flierprops = group_flierprops, \n",
    "              stat_pairs=mannwhitu_testpairs)\n",
    "\n",
    "# Save individual gene plots\n",
    "for idx, gene in enumerate(genes_to_plot):\n",
    "    extent = axs[idx].get_window_extent().transformed(fig.dpi_scale_trans.inverted())\n",
    "    variable = gene\n",
    "    gene_name = variable.split('_')[0]\n",
    "    fig.savefig(f'{column2}_{gene_name}_subplot.png', format='png', dpi=600, bbox_inches=extent.expanded(1.18, 1.2))\n",
    "\n",
    "# Save combined plot\n",
    "plt.savefig(f'{column2}_plot.png', format='png', dpi=600, bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "473d9ea9-d44c-4f03-9047-0cf4c5db0961",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Plots for each Tumor Stage Group with Kruskal Wallis Test Results Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3fb9b0b-0185-43e3-9174-84ddba76fcc3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "column = \"TUMOR_STAGE_T2b\"\n",
    "group = data[column].unique()\n",
    "# Check group[0] is the lower group one, otherwise change to group[1]\n",
    "#group1 = data[data[column] == group[0]]\n",
    "#group1.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003307e6-3c1c-4e6d-830d-e2e2cb0a02ed",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "################################# Kruskal Wallis Plots function:\n",
    "\n",
    "# Import Libraries\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "from statannotations.Annotator import Annotator\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import combinations\n",
    "\n",
    "# Define the plot for Kruskal Wallis Annotations to Group 1\n",
    "def plot_gene_group(data, gene_name, ax, group_num):\n",
    "    # Subset data\n",
    "    column1 = \"TIME_POINT\"\n",
    "    column2 = \"TUMOR_STAGE_T2b\"\n",
    "    group = data[column2].unique()\n",
    "    if group_num == 1:\n",
    "        group_analysis = data[data[column2] == group[0]]\n",
    "        custom_palette = {\"PRE-RT\": \"#E8E8E8\", \"4WKSPD1\": \"#E8E8E8\", \"8WKSPD1\": \"#E8E8E8\"}\n",
    "    elif group_num == 2:\n",
    "        group_analysis = data[data[column2] == group[1]]\n",
    "        custom_palette = {\"PRE-RT\": \"#505050\", \"4WKSPD1\": \"#505050\", \"8WKSPD1\": \"#505050\"}\n",
    "    \n",
    "    PROPS = {'boxprops': {'edgecolor': 'black'}, 'whiskerprops': {'color': 'black'}, 'capprops': {'color': 'black'}}\n",
    "    flierprops = dict(markerfacecolor='grey', markersize=5, markeredgecolor='black')\n",
    "    tp_order = [\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"]\n",
    "    \n",
    "    # Define Kruskal Wallis Test pairs\n",
    "    tps = list([\"PRE-RT\", \"4WKSPD1\", \"8WKSPD1\"])\n",
    "    tp_pairs = list(combinations(tps, r=2))\n",
    "    \n",
    "    # Start Plotting\n",
    "    sns.set(style=\"whitegrid\")\n",
    "    \n",
    "    sns.boxplot(x=column1, y=gene_name, data=group_analysis, order=tp_order, palette=custom_palette, showmeans=False,\n",
    "                showfliers=False, meanprops={\"marker\": \"s\", \"markerfacecolor\": \"black\", \"markersize\": 5, \"markeredgecolor\": \"black\"},\n",
    "                linewidth=1.5, medianprops=dict(color=\"black\"), flierprops=flierprops, **PROPS, ax=ax)\n",
    "\n",
    "    sns.swarmplot(x=column1, y=gene_name, data=group_analysis, order=tp_order, palette=custom_palette, dodge=True,\n",
    "                  size=5, linewidth=0.5, edgecolor=\"black\", ax=ax)\n",
    "\n",
    "    annotator = Annotator(ax=ax, pairs=tp_pairs, data=group_analysis, x=column1, y=gene_name, order=tp_order, palette=custom_palette)\n",
    "    annotator.configure(test='Kruskal', verbose = True).apply_and_annotate() #comparisons_correction=\"Holm-Bonferroni\"\n",
    "\n",
    "    labels = [item.get_text() for item in ax.get_xticklabels()]\n",
    "    labels[0] = 'PRE-PT'\n",
    "    labels[1] = '4WKSPD1'\n",
    "    labels[2] = '8WKSPD1'\n",
    "    ax.set_xticklabels(labels, family=\"serif\")\n",
    "    ax.set_yticklabels(ax.get_yticklabels(), family=\"serif\")\n",
    "\n",
    "    ax.xaxis.tick_top()\n",
    "    ax.xaxis.tick_bottom()\n",
    "    ax.tick_params(axis='x', labelsize=9, labelrotation=0)\n",
    "    ax.tick_params(axis='y', labelsize=9)\n",
    "\n",
    "    ax.set_xlabel('Time Point', family=\"serif\", fontsize=12)\n",
    "    ax.set_ylabel('Gene Expression Fold Change', family=\"serif\", fontsize=12)\n",
    "    variable = gene_name\n",
    "    result = variable.split('_')[0]\n",
    "    ax.set_title(result, family=\"serif\")\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cb2380b-3d34-4c4f-ac48-a866c327f549",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "############################################## Group 1 Plotting\n",
    "\n",
    "# Import Libraries \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statannotations.Annotator import Annotator\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import combinations\n",
    "# Load the data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "# Begin Plotting\n",
    "sns.set(style=\"whitegrid\")\n",
    "fig, axs = plt.subplots(nrows=1, ncols=3, figsize=(30, 8), dpi=600)\n",
    "# Call the function for each gene\n",
    "genes_to_plot = ['MIR17_EXP', 'MIR20B_EXP', 'MIR106A_EXP']  # Add more genes as needed\n",
    "for idx, gene in enumerate(genes_to_plot):\n",
    "    plot_gene_group(data=data, gene_name=gene, ax=axs[idx], group_num=1)\n",
    "\n",
    "# Save individual gene plots\n",
    "name_phenotype = \"TUMORSTAGE_T2b\"\n",
    "\n",
    "#for idx, gene in enumerate(genes_to_plot):\n",
    "    #extent = axs[idx].get_window_extent().transformed(fig.dpi_scale_trans.inverted())\n",
    "    #variable = gene\n",
    "    #gene_name = variable.split('_')[0]\n",
    "    #fig.savefig(f'Less_{name_phenotype}_{gene_name}_subplot.png', format='png', dpi=600, bbox_inches=extent.expanded(1.18, 1.2))\n",
    "\n",
    "# Save combined plot\n",
    "#plt.savefig(f'Less_{name_phenotype}_plot.png', format='png', dpi=600, bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74fec655-e937-40a9-99cb-c293ffea8eb5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "############################################## Group 2 Plotting\n",
    "\n",
    "# Import Libraries \n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from statannotations.Annotator import Annotator\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from itertools import combinations\n",
    "\n",
    "# Load the data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "# Begin Plotting\n",
    "sns.set(style=\"whitegrid\")\n",
    "fig, axs = plt.subplots(nrows=1, ncols=3, figsize=(30, 8), dpi=600)\n",
    "# Call the function for each gene\n",
    "genes_to_plot = ['MIR17_EXP', 'MIR20B_EXP', 'MIR106A_EXP']  # Add more genes as needed\n",
    "for idx, gene in enumerate(genes_to_plot):\n",
    "    plot_gene_group(data=data, gene_name=gene, ax=axs[idx], group_num=2)\n",
    "\n",
    "# Save individual gene plots\n",
    "name_phenotype = \"TUMORSTAGE_T2b\"\n",
    "#for idx, gene in enumerate(genes_to_plot):\n",
    "    #extent = axs[idx].get_window_extent().transformed(fig.dpi_scale_trans.inverted())\n",
    "    #variable = gene\n",
    "    #gene_name = variable.split('_')[0]\n",
    "    #fig.savefig(f'Greater_{name_phenotype}_{gene_name}_subplot.png', format='png', dpi=600, bbox_inches=extent.expanded(1.18, 1.2))\n",
    "\n",
    "# Save combined plot\n",
    "#plt.savefig(f'Greater_{name_phenotype}_plot.png', format='png', dpi=600, bbox_inches='tight')\n",
    "plt.show()\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd38754a-2038-4ccd-b111-72a813ef8cda",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Running Kruskal Wallis Test Manually "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d3e05f3-79d5-45ef-ae3f-b6e7aab14b73",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################## Kruskal-Wallis Results\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "from scipy import stats \n",
    "from scikit_posthocs import posthoc_dunn \n",
    "\n",
    "# Define Kruskal Wallis Test \n",
    "def run_kruskal_group1(dataframe, miRNA_col, group_column, tp_col):\n",
    "    groups = data[group_column].unique()\n",
    "    group1_subset = data[data[group_column] == groups[0]]\n",
    "    tp = np.unique(data[tp_col])\n",
    "    subset_data = [group1_subset[group1_subset[tp_col] == t][miRNA_col] for t in tp]\n",
    "    kruskwall, p_value = stats.kruskal(*subset_data)\n",
    "    print(f\"Kruskal-Wallis Test Result for {miRNA_col} {groups[0]}\")\n",
    "    print(f'Test Statistic: {round(kruskwall, 3)}')\n",
    "    print(f'p-value: {round(p_value, 3)}')\n",
    "    dunn_df = posthoc_dunn(group1_subset, val_col=miRNA_col, group_col=tp_col)\n",
    "    print(f\"Dunn's Test Result for {miRNA_col}, Group: {groups[0]}\")\n",
    "    print(f'{round(dunn_df, 4)}')\n",
    "\n",
    "\n",
    "# Import the Data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "column1 = \"TIME_POINT\"\n",
    "column2 = \"TUMOR_STAGE_T2b\"\n",
    "genes = [\"MIR17_EXP\", \"MIR20B_EXP\", \"MIR106A_EXP\"]\n",
    "\n",
    "for miRNA in genes:\n",
    "    run_kruskal_group1(dataframe=data, miRNA_col=miRNA, group_column=column2, tp_col=column1)\n",
    "    print(\"\\n\")\n",
    "\n",
    "###############################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "031aa51b-98cf-46c3-9d75-76dc14564a83",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "######################################## Kruskal-Wallis Results\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np \n",
    "from scipy import stats \n",
    "from scikit_posthocs import posthoc_dunn \n",
    "\n",
    "# Define Kruskal Wallis Test \n",
    "def run_kruskal_group2(dataframe, miRNA_col, group_column, tp_col):\n",
    "    groups = data[group_column].unique()\n",
    "    group2_subset = data[data[group_column] == groups[1]]\n",
    "    tp = np.unique(data[tp_col])\n",
    "    subset_data = [group2_subset[group2_subset[tp_col] == t][miRNA_col] for t in tp]\n",
    "    kruskwall, p_value = stats.kruskal(*subset_data)\n",
    "    print(f\"Kruskal-Wallis Test Result for {miRNA_col} {groups[1]}\")\n",
    "    print(f'Test Statistic: {round(kruskwall, 4)}')\n",
    "    print(f'p-value: {round(p_value, 4)}')\n",
    "    dunn_df = posthoc_dunn(group2_subset, val_col=miRNA_col, group_col=tp_col)\n",
    "    print(f\"Dunn's Test Result for {miRNA_col}, Group: {groups[1]}\")\n",
    "    print(f'{round(dunn_df, 4)}')\n",
    "\n",
    "\n",
    "# Import the Data\n",
    "data = pd.read_excel(\"/home/johnny.velasquez/OTP_RADGEN_CLINICALDATA_ANALYSIS/DATA_PREPROCESSING/OTP_RADIOGENOMICS_FINAL_CLINICALDATASET_DEC12_2023_V4.xlsx\")\n",
    "\n",
    "column1 = \"TIME_POINT\"\n",
    "column2 = \"TUMOR_STAGE_T2b\"\n",
    "genes = [\"MIR17_EXP\", \"MIR20B_EXP\", \"MIR106A_EXP\"]\n",
    "\n",
    "for miRNA in genes:\n",
    "    run_kruskal_group2(dataframe=data, miRNA_col=miRNA, group_column=column2, tp_col=column1)\n",
    "    print(\"\\n\")\n",
    "\n",
    "###############################################################"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "UFRC Python-3.10",
   "language": "python",
   "name": "python3-3.10-ufrc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
